apps/test/ChartApiTest.js

```diff
@@ -0,0 +1,367 @@
+      chartApi.drawChartFromRecords('missingId')
+          .then(onResolve, onReject)
+          .then(ensureDone(testDone, function () {
+            assert.equal(
+                rejection.message,
+                'Unable to render chart into element "missingId".');
+          }));
+    });
+
+    it ("rejects if element is wrong type", function (testDone) {
+      chartApi.drawChartFromRecords('fakeImg')
+          .then(onResolve, onReject)
+          .then(ensureDone(testDone, function () {
+            assert.equal(
+                rejection.message,
+                'Unable to render chart into element "fakeImg".');
+          }));
+    });
+
+    it ("rejects if chart type is not supported", function (testDone) {
+      chartApi.drawChartFromRecords('fakeDiv', 'badType')
+          .then(onResolve, onReject)
+          .then(ensureDone(testDone, function () {
+            assert.equal(
+                rejection.message,
+                'Unsupported chart type "badType".');
+          }));
+    });
+
+    it ("rejects if no columns array provided", function (testDone) {
+      chartApi.drawChartFromRecords('fakeDiv', ChartType.PIE, 'fakeTable')
+          .then(onResolve, onReject)
+          .then(ensureDone(testDone, function () {
+            assert.equal(
+                rejection.message,
+                "Not enough columns for chart; expected at least 2.");
+          }));
+    });
+
+    it ("rejects if zero columns provided", function (testDone) {
+      chartApi.drawChartFromRecords('fakeDiv', ChartType.PIE, 'fakeTable', [])
+          .then(onResolve, onReject)
+          .then(ensureDone(testDone, function () {
+            assert.equal(
+                rejection.message,
+                'Not enough columns for chart; expected at least 2.');
+          }));
+    });
+
+    it ("rejects if only one column provided", function (testDone) {
+      chartApi.drawChartFromRecords('fakeDiv', ChartType.PIE, 'fakeTable', ['column1'])
+          .then(onResolve, onReject)
+          .then(ensureDone(testDone, function () {
+            assert.equal(
+                rejection.message,
+                'Not enough columns for chart; expected at least 2.');
+          }));
+    });
+
+    it ("when fulfilled, makes API calls", function (testDone) {
+      chartApi.drawChartFromRecords('fakeDiv', ChartType.PIE, 'fakeTable', ['column1', 'column2'])
+          .then(onResolve, onReject)
+          .then(ensureDone(testDone, function () {
+            assert.isNull(rejection);
+          }));
+    });
+
+    it ("warns about empty dataset", function (testDone) {
+      chartApi.drawChartFromRecords('fakeDiv', ChartType.PIE, 'fakeTable',
+          ['column1', 'column2'])
+          .then(onResolve, onReject)
+          .then(ensureDone(testDone, function () {
+            assertWarns(chartApi, /No data\./);
+          }));
+    });
+
+    it ("does not warn about empty dataset when given data", function (testDone) {
```

Comment:
How about adding a similar test to this one but without the `columns` parameter, which does `assertNotWarns(..., "Not enough columns...")` ? To verify that columns can be inferred.